File: Golang/1-50/14.go
package main

import (
	"bufio"
	"fmt"
	"math/big"
	"os"
	"strconv"
	"strings"
)

func modularPow(baseNumber, exponent, modulus int64) int64 {
	result := big.NewInt(1)
	base := big.NewInt(baseNumber)
	mod := big.NewInt(modulus)

	for exponent > 0 {
		if exponent%2 == 1 {
			result.Mul(result, base).Mod(result, mod)
		}
		base.Mul(base, base).Mod(base, mod)
		exponent /= 2
	}

	return result.Int64()
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	inputs := strings.Fields(input)
	n, _ := strconv.ParseInt(inputs[0], 10, 64)
	k, _ := strconv.ParseInt(inputs[1], 10, 64)

	var output int64
	if n == 0 {
		output = 1
	} else {
		output = modularPow(k+1, n, 1000000007)
	}

	fmt.Println(output)
}

File: C sharp/1-50/14.cs
using System;

class Program
{
    static long ModularPow(long baseNumber, long exponent, long modulus)
    {
        long result = 1;
        while (exponent > 0)
        {
            if (exponent % 2 == 1)
            {
                result = (result * baseNumber) % modulus;
            }
            baseNumber = (baseNumber * baseNumber) % modulus;
            exponent /= 2;
        }
        return result;
    }

    static void Main()
    {
        string[] inputs = Console.ReadLine().Split();
        int n = int.Parse(inputs[0]);
        int k = int.Parse(inputs[1]);

        long output;
        if (n == 0)
        {
            output = 1;
        }
        else
        {
            output = ModularPow(k + 1, n, 1000000007);
        }

        Console.WriteLine(output);
    }
}
